extract code from print_cfl into dss_hvtensor routine

mt5555 opened the pull request:
HOMME fortran init code refactor: important code hidden in print_cfl() is extracted into a new routine, dss_hvtensor

dss_hvtensor improves HV results by making the HV coefficients smooth accross elements.  It also does a bilinear projection of the coefficients within each element to minimize oscillations in the tensor coefficieints.

[BFB]

rljacob commented:
@tcclevenger I think this is ready

tcclevenger commented:
I'll merge to next

tcclevenger commented:
@rljacob Currently trying to setup homme standalone tests to verify this fix. Will merge to next once I complete that.

tcclevenger commented:
Merged to next

tcclevenger commented:
I suspect this PR could be the reason for all the testing DIFFs. Reverting from next to investigate.

mt5555 commented:
one guess:  keep the new routine in the original file, global_norms_mod.F90, so it will be compiled with the same optimization.

tcclevenger commented:
> one guess: keep the new routine in the original file, global_norms_mod.F90, so it will be compiled with the same optimization.

Normalized errors seem to be too high for just optimization differences. I'm seeing them as large as 2e+1 in some cases.

tcclevenger commented:
> @tcclevenger Any new info?

Not yet, I'm going to carve out time next week to solve this.

rljacob commented:
@tcclevenger any update?

tcclevenger commented:
> @tcclevenger any update?

Looked at this last week and it not obvious whats going wrong. I'm able to reproduce the error, hoping to have some time in the next week or two to figure it out.

tcclevenger commented:
@mt5555 You were correct, it was fixed by keeping the impl in the same file. Looking back through the CDASH from when I merged to next, it looks to be only intel compiler that triggered the bug. An example of DIFFs I saw in `SMS.ne4pg2_oQU480.F2010.pm-cpu_intel.eam-thetanh_ftype1`
```
1636  x2l_Faxa_snowc   (doml_nx,doml_ny,time)  t_index =      1     1
1637          27      384  (   341,     1,     1) (     3,     1,     1) (   341,     1,     1) (    53,     1,     1)
1638                  201   3.503695987285971E-05   0.000000000000000E+00 3.5E-05  3.503695987285971E-05 9.9E-02  4.259984908567328-110
1639                  201   1.829463226967772E-05   0.000000000000000E+00          0.000000000000000E+00          0.000000000000000E+00
1640                  384  (   374,     1,     1) (     3,     1,     1)
1641           avg abs field values:    2.751477713696158E-07    rms diff: 2.6E-06   avg rel diff(npos):  9.9E-02
1642                                    1.347734549979935E-07                        avg decimal digits(ndif):  0.3 worst:  0.0
1643  RMS x2l_Faxa_snowc                   2.5518E-06            NORMALIZED  1.2450E+01

1681  x2l_Faxa_swndf   (doml_nx,doml_ny,time)  t_index =      1     1
1682          77      384  (   278,     1,     1) (     3,     1,     1) (    77,     1,     1) (    77,     1,     1)
1683                  201   1.939413049475074E+02   0.000000000000000E+00 1.6E+02  1.701407560647017E+02 6.9E-02  1.701407560647017E+02
1684                  201   1.946504984163617E+02   0.000000000000000E+00          1.282623020327691E+01          1.282623020327691E+01
1685                  384  (   130,     1,     1) (     3,     1,     1)
1686           avg abs field values:    1.478772437679993E+01    rms diff: 1.7E+01   avg rel diff(npos):  6.9E-02
1687                                    1.320859658869269E+01                        avg decimal digits(ndif):  1.1 worst:  0.0
1688  RMS x2l_Faxa_swndf                   1.7031E+01            NORMALIZED  1.2167E+00
```

Should we be concerned about diffs this large just because of a file change?

If not, I'll merge to next today and create a PR adding `dss_hvtensor()` call to EAMxx.

mt5555 commented:
So for now - you put the routine back in the original file (but kept the modularization) and it's now BFB?

Roundoff level changes (such as from optimization changes which can depend on the file) do grow exponentially.  So by just looking at the numbers above, not possible to confirm if that is acceptable growth or not.  

Can we be lazy and just leave this routine in global_norms_mod.F90 (and rely on BFB testing)?  

If we really want to move it out of this routine, lets do a second PR that only moves the subroutine (so it will be obvious just from inspection that the change is roundoff and doesn't need more sophisticated testing).

tcclevenger commented:
> Can we be lazy and just leave this routine in global_norms_mod.F90 (and rely on BFB testing)?

For sure, PM-CPU didn't report on cdash so may have to wait a day, but otherwise looks BFB.

rljacob commented:
This was placed on next yesterday (with the wrong PR number) and its causing testing fails on an odd assortment of cases.

tcclevenger commented:
Reverted from next while I debug.

rljacob commented:
Found a different error in a debugging case.

 46: forrtl: severe (408): fort: (7): Attempt to use pointer DESC when it is not associated with a target
 46:
 46: Image              PC                Routine            Line        Source
 46: libpnetcdf.so.3.0  0000155554F111A2  for_emit_diagnost     Unknown  Unknown
 46: e3sm.exe           00000000041C8B23  edge_mod_base_mp_         413  edge_mod_base.F90
 46: e3sm.exe           000000000431FD27  global_norms_mod_         444  global_norms_mod.F90
 46: e3sm.exe           000000000443F683  prim_driver_base_         986  prim_driver_base.F90
 46: e3sm.exe           0000000005C2FF83  dyn_comp_mp_dyn_i         380  dyn_comp.F90
 46: libiomp5.so        00001555531443F3  __kmp_invoke_micr     Unknown  Unknown
 46: libiomp5.so        00001555530C8273  Unknown               Unknown  Unknown
 46: libiomp5.so        00001555530C721E  Unknown               Unknown  Unknown
 46: libiomp5.so        00001555531448CC  Unknown               Unknown  Unknown
 46: libpthread-2.28.s  00001555483EA1CA  Unknown               Unknown  Unknown
 46: libc-2.28.so       0000155548056E73  clone                 Unknown  Unknown
 46: [chr-0506:2841896:0:2842014] Caught signal 11 (Segmentation fault: address not mapped to object at address (nil))

cases:
SMS_D_Ln5.conusx4v1pg2_r05_IcoswISC30E3r5.F2010.chrysalis_intel

tcclevenger committed:
extract code from print_cfl into dss_hvtensor routine

HOMME fortran init code refactor: important code hidden in print_cfl()
is extracted into a new routine, dss_hvtensor

dss_hvtensor improves HV results by making the HV coefficients
smooth accross elements.  It also does a bilinear projection
of the coefficients within each element to minimize oscillations
in the tensor coefficieints.

tcclevenger committed:
EdgeBuffer cleanup

Remove unused EdgeBuffer types from test_global_integrals and add
comment about why edgebuf should be module static

mt5555 reviewed (commented):
looks good.  why is the "approve" button commented out for me?

tcclevenger commented:
> looks good. why is the "approve" button commented out for me?

It's your PR. I guess we need a 3rd opinion?

rljacob commented:
Its clear to merge because it already has one review.  Feel free to ask for more if you want.

tcclevenger commented:
Merged to next.

tcclevenger commented:
@rljacob `e3sm_prod_next_intel` never reported for PM-CPU. Other than that the dash board looks clean wrt this PR. Am I good to merge? Or should I wait for all of next testing to report?

rljacob commented:
Yes go ahead.  The chrysalis report on integration is fine.
